### ISSUE TITLE
Brittle guidance regarding idempotent activities

### ISSUE THREAD

**Issue by Tamschi** [w3c-social/activitypump#485]
This is regarding

- https://www.w3.org/wiki/index.php?title=ActivityPub/Primer/Follow_activity&oldid=116221#Multiple_Follow_activities
- https://www.w3.org/wiki/index.php?title=ActivityPub/Primer/Like_activity&oldid=116220#Multiple_Like_activities

(current versions as of writing).

If only the oldest activity of an idempotent kind is stored, it's relatively easy to (due to very unreliable network conditions or [other bugs](https://github.com/snarfed/bridgy-fed/issues/1361)) arrive at a state where the sender is permanently unable to directly remove such a relation because their instance doesn't allow them to re-send an `Undo`-activity correctly targeting the activity that originally established the relation.

To mitigate this,

1. the receiving server of subjectively superfluous `Follow` or `Like` activities should[^1] replace the activity `@id` they have on file for the existing relation with the latest one that would have created it while still cancelling any other side-effects of the latter.
2. If a stored `Undo` targeting this new `Follow` or `Like` has already been received and buffered, its effects should be applied (as if) after updating the stored `@id`.

[^1]: arguably "SHOULD", but I'm not sure if those keywords are considered appropriate for use in the Primer

**Comment by Tamschi** [comment]
This was originally discussed in https://github.com/w3c/activitypub/issues/381.

(Thanks @snarfed for pointing this out! I tried searching for it but must have just barely missed the keywords.)

**Comment by evanp** [comment]
I agree that it probably makes sense for maximum interoperability for the receiving server to store not only the first Follow or Like activity, but to store additional Follow or Like activities, primarily for Undo. It is possible to ignore the `id` of an activity being undone, and just execute based on shape, but there are probably some security issues there. I'm going to mark this ticket as requiring an update to the primer, in order to suggest maximizing Undo functionality.

**Comment by nightpool** [comment]
what security issues?
